### ISSUE TITLE
Use of resource "aws_s3_bucket_object" with etag = filemd5() function results in large memory usage

### ISSUE THREAD

**Issue by davidcallen** [hashicorp/terraform-provider-aws#11656]
Running a typical upload of file to S3 results in large memory usage when file size is large. Example below : 

```
resource "aws_s3_bucket_object" "object" {
  bucket = "your_bucket_name"
  key    = "new_object_key"
  source = "path/to/file"
  etag = "${filemd5("path/to/file")}"
}
```

When uploading a large **file of 3.5GB** the terraform process increased in memory from the typical 85MB (resident set size) up **to 4GB** (resident set size). The memory size remains high even when waiting at the "apply changes" prompt.

It looks like the use of **filemd5()** function is generating the md5 checksum by loading the entire file into memory and **then not releasing that memory** after finishing.  

The problem is increased when processing multiple files (using a fileset() function for a for loop) e.g. : 
```
resource "aws_s3_bucket_object" "files-store-ibm-ilmt-bigfix" {
  for_each = fileset("${path.module}/files-for-uploading/", "*")
  bucket  = "your_bucket_name"
  key     = "files-uploaded/${each.value}"
  source  = each.value
  etag    = "${filemd5("${path.module}/files-for-uploading/${each.value}")}"
}
```

**Environment :** 
Fedora Linux 29 on a 8GB RAM laptop (runs out of RAM easily with this issue).

```
terraform -v
Terraform v0.12.13
provider.aws v2.45.0
provider.null v2.1.2
```

**Expected Outcome :**
Create the md5 checksum reading file in buffers of sensible size.  I tried running the "md5sum" utility under linux (fedora 29) on the 3.5GB file and this 1.3MB  (resident set size).  That would be a good example of how to implement the checksum.

**Workaround :** 
My current workaround for this may help others with problem. Its not ideal in that it is using a shell command but works well for now. Use alternative below : 

```
resource "null_resource" "files-store-upload" {
  triggers = {
    always_run = "${timestamp()}"
  }
  provisioner "local-exec" {
    command = "aws s3 sync ${path.module}/files-for-uploading/ s3://${aws_s3_bucket.files-store.id}/uploaded-files/"
  }
}
```

**Comment by ewbankkit** [comment]
@davidcallen I suggest raising an issue in the [Terraform core repository](https://github.com/hashicorp/terraform/issues) as that's where `filemd5()` is implemented and linking to this issue.
It seem like a "streaming" implementation could be used, e.g. https://tip.golang.org/pkg/crypto/md5/#example_New_file.

**Comment by ewbankkit** [comment]
Related:
* https://github.com/terraform-providers/terraform-provider-aws/issues/788

**Comment by davidcallen** [comment]
Thanks @ewbankkit for the fast triage and response.  
Good point that the issue is probably all in filemd5() function and therefore I have created issue in Terraform core :  https://github.com/hashicorp/terraform/issues/23890

If I get time to learn some golang I'll look at finding a fix for it.

**Comment by ewbankkit** [comment]
@davidcallen Improvements to the Terraform file hashing functions (including `filemd5`) were made in https://github.com/hashicorp/terraform/pull/28681 and released in Terraform [**v0.15.4**](https://github.com/hashicorp/terraform/blob/v0.15/CHANGELOG.md#0154-may-19-2021).
Could you please upgrade to that version of the Terraform CLI, if possible, and retry your scenario?

**Comment by davidcallen** [comment]
Hi @ewbankkit - I have just upgraded to v0.15.4 and the fix works :).  Nice one. Many thanks
